File: C++/0950.cpp
#include <vector>
#include <algorithm>
#include <queue>

class Solution {
public:
    std::vector<int> deckRevealedIncreasing(std::vector<int>& deck) {
        std::sort(deck.begin(), deck.end());
        std::vector<int> ans(deck.size());
        std::queue<int> idx;
        
        for(int i = 0; i < deck.size(); i++){
            idx.push(i);
        }
        for(int card : deck){
            ans[idx.front()] = card;
            idx.pop();
            if(!idx.empty()){
                idx.push(idx.front());
                idx.pop();
            }
        }
        return ans;
    }
};

// Topic: Array
// Time Complexity: O(nlogn)
// Space Complexity: O(n)

File: Python/0950.py
from typing import List
from collections import deque

class Solution:
    def deckRevealedIncreasing(self, deck: List[int]) -> List[int]:
        deck.sort()
        n = len(deck)
        ans = [0] * n
        idx = deque(range(n))
        for card in deck:
            ans[idx.popleft()] = card
            if idx:
                idx.append(idx.popleft())
                
        return ans

# Topic: Array
# Time Complexity: O(nlogn)
# Space Complexity: O(n)
